### PR TITLE
feat(langgraph): add deterministic CI failure summarizer spike

### PR DESCRIPTION
**Pull request by kami619**
## Description
  Add a new `langgraph` spike template, `ci_failure_summarizer`, for daily QG4 failure triage on OpenShift.
  This PR:
  - adds a new standard agent template at `agents/langgraph/templates/ci_failure_summarizer`
  - ingests GitHub Actions workflow/job data for the QG4 workflow and groups failures deterministically
  - extracts failed-step log evidence, persists typed failure evidence in PostgreSQL, and renders `/summarize` output deterministically (no LLM used in the summarization path)
  - posts a second top-level Slack triage summary via the existing incoming webhook model
  - adds an OpenShift CronJob example for running `/summarize` daily at **11 AM IST**
  - validates the end-to-end path in `ci-testing`, including authenticated log access, Slack posting, and manual CronJob smoke execution
  Motivation:
  - prove a lightweight, cluster-hosted CI summarizer can reuse existing services and produce actionable summaries without relying on heuristic LLM diagnosis
  - establish a reusable deterministic evidence pipeline before layering on optional LLM-assisted enhancements later
  ## Jira Ticket
  `RHAIENG-6286`
  ## Testing
  - [x] `make test` passes (run from the affected agent directory)
  - [x] Manual testing performed (describe steps below)
  - [ ] No testing required (documentation/config change only)
  Additional verification:
  - `make test` in `agents/langgraph/templates/ci_failure_summarizer` -> `61 passed`
  - deployed to `ci-testing` and verified `GET /health`
  - verified authenticated `POST /summarize` against QG4 run `29716559471`
  - verified Slack posting from the summarizer service
  - applied `examples/trigger_daily_summary_cronjob.yaml`
  - created a manual Job from the CronJob and verified successful completion plus `/summarize` execution
  ## Checklist
  - [ ] I have read [CONTRIBUTING.md](CONTRIBUTING.md)
  - [x] No `.env` or secret files are included in this PR
  - [x] All changes are within scope of the linked Jira ticket (if not, explain in Description)
  ## New Agent Proposal
  - [ ] Fit check completed ([how to run](CONTRIBUTING.md#proposing-a-new-agent))
  - Fit check discussion: `TBD / not run yet`
  - Fit score: `TBD`
  ## Review Guidance
  Please focus on:
  - `src/ci_failure_summarizer/failure_evidence.py` for the shared typed evidence model
  - `src/ci_failure_summarizer/github_client.py` for failed-step log extraction
  - `src/ci_failure_summarizer/summary_composer.py` for deterministic summary rendering
  - `src/ci_failure_summarizer/orchestrator.py` for persistence + summary-history wiring
  - `examples/trigger_daily_summary_cronjob.yaml` as an example operator-managed scheduler, not a Helm-managed resource
  Design notes:
  - `/summarize` is intentionally deterministic in this PR; the inherited `/chat/completions` path remains from the base template, but CI triage does not depend on an LLM
  - evidence is persisted on incidents and a minimal audit snapshot is stored in summary history
  - the CronJob example targets **11 AM IST** using `timeZone: Asia/Kolkata`
  ## Related PRs
  None

### CODE REVIEW

**Comment by kami619** [comment]
@coderabbit I have fixed all your previous review items, can you take another look ?

**Comment by kami619** [comment]
@coderabbit can you take another look at your comments and the latest commit based fixes ?

**Comment by kami619** [comment]
@mpk-droid this should be in a state to start the review and the flow itself is working based on the deployment done from this branch on the team cluster.

File: agents/langgraph/templates/ci_failure_summarizer/src/ci_failure_summarizer/incident_store.py [hunk 1]
```diff
@@ -0,0 +1,277 @@
...
+                )
+            else:
+                conn.execute(
+                    """
+                    DELETE FROM ci_slack_post_claims
+                    WHERE run_id = %s AND posted_at IS NULL
+                    """,
+                    (run_id,),
+                )
+            conn.commit()
```
[the remark this answers is not in the document]
Comment (kami619) [reply agents/langgraph/templates/ci_failure_summarizer/src/ci_failure_summarizer/incident_store.py]: won't do this iteration.

File: agents/langgraph/templates/ci_failure_summarizer/src/ci_failure_summarizer/github_client.py [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+        payload = self._request("GET", path, params={"per_page": 1}).json()
+        runs = payload.get("workflow_runs") or []
+        if not runs:
+            return None
+        return WorkflowRun.from_api(runs[0])
+
+    def list_jobs(self, run_id: int, *, per_page: int = 100) -> list[WorkflowJob]:
+        path = f"/repos/{self.repository}/actions/runs/{run_id}/jobs"
+        # Spike assumption: QG4 runs have <=100 jobs; no Link pagination yet.
+        payload = self._request("GET", path, params={"per_page": per_page}).json()
```
Comment (mpk-droid) [inline agents/langgraph/templates/ci_failure_summarizer/src/ci_failure_summarizer/github_client.py:178]: **nit:** `per_page=100` with no pagination — if the job count ever exceeds 100, results are silently truncated (no error, no warning). Not urgent today (~15 agents), but worth adding a log warning when the response count equals `per_page` so truncation doesn't go unnoticed as the repo grows.

Assisted by Claude Opus 4.6 (1M context)

File: agents/langgraph/templates/ci_failure_summarizer/src/ci_failure_summarizer/github_client.py [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+
+    def get_run(self, run_id: int) -> WorkflowRun:
+        path = f"/repos/{self.repository}/actions/runs/{run_id}"
+        payload = self._request("GET", path).json()
+        return WorkflowRun.from_api(payload)
+
+    def get_latest_run(self, workflow_file: str) -> WorkflowRun | None:
+        workflow_ref = workflow_api_ref(workflow_file)
+        path = f"/repos/{self.repository}/actions/workflows/{workflow_ref}/runs"
+        payload = self._request("GET", path, params={"per_page": 1}).json()
```
Comment (mpk-droid) [inline agents/langgraph/templates/ci_failure_summarizer/src/ci_failure_summarizer/github_client.py:169]: **nit:** `get_latest_run` fetches only the single most recent workflow run (`per_page: 1`). Currently we have one nightly QG4 run so this is fine, but if we ever add more frequent runs (or re-runs), any runs between CronJob ticks would be skipped silently.

This is mainly a limitation of the CronJob-based trigger model — it assumes exactly one run happened since the last tick. A trigger-based approach (e.g., a post-failure step in the QG4 workflow itself calling `/summarize?run_id=$GITHUB_RUN_ID`) would eliminate this entirely and also resolve the schedule timing issue in the CronJob example.

Assisted by Claude Opus 4.6 (1M context)

File: agents/langgraph/templates/ci_failure_summarizer/src/ci_failure_summarizer/github_client.py (hunk 2 above)
Comment (kami619) [reply agents/langgraph/templates/ci_failure_summarizer/src/ci_failure_summarizer/github_client.py:178]: good point, I will add that suggestion.

File: agents/langgraph/templates/ci_failure_summarizer/src/ci_failure_summarizer/github_client.py (hunk 3 above)
Comment (kami619) [reply agents/langgraph/templates/ci_failure_summarizer/src/ci_failure_summarizer/github_client.py:169]: that is correct and your intuition is right about having this as a push model instead of pull would be ideal and having a step at the end of each of the workflow would make this more scalable. However the look-up logic to see all the jobs will still remain as individual `/summarize` calls would only make this execute and provide for each failure run separately but won't be a catch all job. I would leave this out of scope for now and we can add a new ticket, WDYT ?

File: agents/langgraph/templates/ci_failure_summarizer/src/ci_failure_summarizer/github_client.py [hunk 4]
```diff
@@ -0,0 +1,235 @@
...
+    """Return the workflow file basename accepted by GitHub workflow endpoints."""
+    return workflow_file.rsplit("/", 1)[-1]
+
+
+class GitHubActionsClient:
+    """Fetch public GitHub Actions metadata; degrade when logs require auth."""
+
+    def __init__(self, repository: str, token: str | None = None) -> None:
+        self.repository = repository
+        self._session = requests.Session()
```
Comment (mpk-droid) [inline agents/langgraph/templates/ci_failure_summarizer/src/ci_failure_summarizer/github_client.py:107]: **nit (production hardening):** `requests.Session` is created here but never closed — each `/summarize` call leaks a session. Similarly, `IncidentStore` opens a fresh psycopg connection per method call (acknowledged in comments). For production, consider using a connection pool for PostgreSQL and closing/reusing the HTTP session.

Assisted by Claude Opus 4.6 (1M context)

File: agents/langgraph/templates/ci_failure_summarizer/src/ci_failure_summarizer/slack_notifier.py [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+    payload: dict[str, Any],
+    timeout: int = 15,
+) -> None:
+    try:
+        response = requests.post(webhook_url, json=payload, timeout=timeout)
+    except requests.RequestException as exc:
+        raise RuntimeError(sanitize_slack_error(exc, webhook_url=webhook_url)) from exc
+    if response.status_code >= 400:
+        raise RuntimeError(
+            f"Slack webhook returned HTTP {response.status_code}: {response.text}"
```
Comment (mpk-droid) [inline agents/langgraph/templates/ci_failure_summarizer/src/ci_failure_summarizer/slack_notifier.py]: **nit:** `response.text` here is included unsanitized in the `RuntimeError`, which surfaces as an HTTP 502 detail to the API caller via `main.py:612`. If Slack's error response echoes back the webhook URL, it would leak to the caller. There's already a `sanitize_slack_error` function in this file — apply it to `response.text` before including it in the exception.

Assisted by Claude Opus 4.6 (1M context)

File: agents/langgraph/templates/ci_failure_summarizer/src/ci_failure_summarizer/incident_store.py [hunk 6]
```diff
@@ -0,0 +1,316 @@
...
+        with psycopg.connect(
+            self.db_uri,
+            row_factory=dict_row,
+            connect_timeout=10,
+        ) as conn:
+            yield conn
+
+    def setup(self) -> None:
+        with self._connection() as conn:
+            conn.execute(SCHEMA_SQL)
```
Comment (mpk-droid) [inline agents/langgraph/templates/ci_failure_summarizer/src/ci_failure_summarizer/incident_store.py:79]: **nit (production hardening):** This loop executes one INSERT...ON CONFLICT per failure — classic N+1 pattern. Fine for spike load (10-20 failures), but for production consider batching with `executemany()` or a multi-row VALUES clause to reduce round-trips.

Assisted by Claude Opus 4.6 (1M context)

File: agents/langgraph/templates/ci_failure_summarizer/src/ci_failure_summarizer/orchestrator.py [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+                workflow_name=run.name,
+                failures=[],
+                incidents=[],
+                summary_text=summary_text,
+                slack_posted=False,
+                slack_skipped_reason="latest run did not fail",
+            )
+
+        jobs = self.github.list_jobs(run.id)
+        log_results = {
```
Comment (mpk-droid) [inline agents/langgraph/templates/ci_failure_summarizer/src/ci_failure_summarizer/orchestrator.py]: **nit (production hardening):** Log fetches are serial here — one HTTP request per failed job, waiting for each before starting the next. With 10 failures at ~150ms each, that's ~1.5s of serial blocking. For production, consider parallelizing with `ThreadPoolExecutor.map()` since each fetch is independent.

Assisted by Claude Opus 4.6 (1M context)

File: agents/langgraph/templates/ci_failure_summarizer/src/ci_failure_summarizer/incident_store.py [hunk 8]
```diff
@@ -0,0 +1,316 @@
...
+            if row is None:
+                raise RuntimeError(
+                    "Failed to record summary history row for CI summarizer run"
+                )
+            return int(row["id"])
+
+    def get_incident(self, fingerprint: str) -> Incident | None:
+        with self._connection() as conn:
+            row = conn.execute(
+                "SELECT * FROM ci_incidents WHERE fingerprint = %s",
```
Comment (mpk-droid) [inline agents/langgraph/templates/ci_failure_summarizer/src/ci_failure_summarizer/incident_store.py:191]: **nit:** This query filters by `run_id` with `ORDER BY created_at DESC, id DESC` but there's no index on `run_id` in the schema. The table grows with every `/summarize` call. Consider adding:
```sql
CREATE INDEX IF NOT EXISTS idx_ci_summary_history_run_id
  ON ci_summary_history(run_id, created_at DESC, id DESC);
```

Assisted by Claude Opus 4.6 (1M context)

File: agents/langgraph/templates/ci_failure_summarizer/main.py [hunk 9]
```diff
@@ -0,0 +1,650 @@
...
+                }
+            ],
+            "context": context_messages,
+            "usage": _extract_usage(new_messages),
+        }
+
+    except Exception:
+        logger.exception("Error processing chat completion request")
+        raise HTTPException(status_code=500, detail="Error processing request")
+
```
Comment (mpk-droid) [inline agents/langgraph/templates/ci_failure_summarizer/main.py:393]: **nit:** This function is ~140 lines mixing DB saver init, agent creation, stream event iteration, three different SSE transformations, and error handling. Consider extracting the event-type-specific chunk builders (`on_chat_model_stream`, `on_chat_model_end`, `on_tool_end`) into small helper functions to make the flow easier to follow and test.

Assisted by Claude Opus 4.6 (1M context)

File: agents/langgraph/templates/ci_failure_summarizer/main.py [hunk 10]
```diff
@@ -0,0 +1,650 @@
...
+    model: str | None = Field(
+        None,
+        description="ID of the model to use. Defaults to the server's configured MODEL_ID.",
+    )
+    stream: bool = Field(
+        False,
+        description="If true, partial message deltas will be sent as SSE `data: {json}\\n\\n` events, terminated by `data: [DONE]\\n\\n`.",
+    )
+    thread_id: str | None = Field(
+        None,
```
Comment (mpk-droid) [inline agents/langgraph/templates/ci_failure_summarizer/main.py:75]: **nit (production hardening):** Any client-supplied `thread_id` is passed directly to the checkpoint saver without ownership checks. An attacker who guesses a thread ID can read or write to arbitrary conversation threads. Low priority since this endpoint is scaffold (not the spike's workflow), but worth a TODO for production.

Assisted by Claude Opus 4.6 (1M context)

**Comment by mpk-droid** [comment]
**nit:** `failure_evidence.py` is the core of the spike — 14 functions for evidence extraction, marker scoring, fingerprint building, and evidence merging — but has no dedicated test file. Only `select_best_error_marker` has a direct unit test; the rest are exercised indirectly through other tests. Given this module is central to the spike's value proposition, it would benefit from its own `tests/test_failure_evidence.py` covering edge cases (empty logs, duplicate markers, signature collisions, metadata round-tripping).

Assisted by Claude Opus 4.6 (1M context)

**Comment by mpk-droid** [comment]
**Question:** The playground UI (`playground/app.py` + `index.html`, ~600 LoC) provides a chat interface for `/chat/completions`, but the spike's workflow is automated (cron/manual trigger → `/summarize` → Slack). Is the playground UI needed for this agent, or can it be removed to reduce the surface area?

Assisted by Claude Opus 4.6 (1M context)

**Comment by mpk-droid** [comment]
**nit:** The agent is called `ci_failure_summarizer` and has `MODEL_ID`/`BASE_URL`/`API_KEY` in `.env.example`. A future reader will naturally assume the summarizer uses an LLM. It doesn't — `/summarize` is purely deterministic (regex matching, fingerprinting, grouping). The LLM config is only for the inherited `/chat/completions` scaffold. Worth calling this out explicitly in the README so someone doesn't spend time hunting for where the LLM call happens in the summarization path.

Assisted by Claude Opus 4.6 (1M context)

**Comment by mpk-droid** [comment]
**Question:** The behavioral test suite (7 files, ~689 LoC) covers latency budgets, memory isolation, streaming parity, and tool usage — all targeting the `/chat/completions` endpoint. Since the spike's workflow is `/summarize` (not chat), are these behavioral tests adding value here, or would the test effort be better invested in unit tests for the actual triage logic (e.g., `failure_evidence.py`)?

Assisted by Claude Opus 4.6 (1M context)

**Comment by kami619** [comment]
@mpk-droid excellent questions, I will take a look and get back to you.

**Comment by kami619** [comment]
> Question: The playground UI (playground/app.py + index.html, ~600 LoC) provides a chat interface for /chat/completions, but the spike's workflow is automated (cron/manual trigger → /summarize → Slack). Is the playground UI needed for this agent, or can it be removed to reduce the surface area?

Great catch @mpk-droid , this is almost certainly dead code that can be removed as I pivoted slightly on the workflow as the existing template that I based this agent off of doesn't support rich text reply. I will remove this and make the docs more water tight.

File: agents/langgraph/templates/ci_failure_summarizer/src/ci_failure_summarizer/incident_store.py (hunk 6 above)
Comment (kami619) [reply agents/langgraph/templates/ci_failure_summarizer/src/ci_failure_summarizer/incident_store.py:79]: I am going to ignore this nit :) this will be subject many other production hardening stories in the client's purview, this showcase will be fine for the template we have in place.

File: agents/langgraph/templates/ci_failure_summarizer/src/ci_failure_summarizer/orchestrator.py (hunk 7 above)
Comment (kami619) [reply agents/langgraph/templates/ci_failure_summarizer/src/ci_failure_summarizer/orchestrator.py]: this is a good suggestion, I am going to take this feedback into implementation loop.

File: agents/langgraph/templates/ci_failure_summarizer/src/ci_failure_summarizer/incident_store.py (hunk 8 above)
Comment (kami619) [reply agents/langgraph/templates/ci_failure_summarizer/src/ci_failure_summarizer/incident_store.py:191]: good suggestion, I shall review the SQL initialization step to add this in, thanks @mpk-droid

File: agents/langgraph/templates/ci_failure_summarizer/main.py (hunk 10 above)
Comment (kami619) [reply agents/langgraph/templates/ci_failure_summarizer/main.py:75]: going to ignore this as there would be many security, performance hardening stories that would need to productize this workflow.

**Comment by kami619** [comment]
@mpk-droid PTAL, i have addressed and implemented the needed feedback at the moment.

File: agents/langgraph/templates/ci_failure_summarizer/playground/app.py [hunk 11]
```diff
@@ -0,0 +1,152 @@
...
+                    "status": "unreachable",
+                    "error": "Agent is unreachable. Please try again later.",
+                }
+            ),
+            503,
+        )
+
+
+@app.route("/api/chat", methods=["POST"])
+def chat():
```
[the remark this answers is not in the document]
Comment (kami619) [reply agents/langgraph/templates/ci_failure_summarizer/playground/app.py:74]: Fixed in `e3a1de1`. Instead of keeping and patching the inherited Flask playground, I removed the playground surface entirely for this spike: deleted `playground/app.py` and `playground/templates/index.html`, removed the `/` and `/images/{filename}` routes from `main.py`, dropped the Docker packaging for those assets, and replaced the old test with route-removal assertions. That keeps the required chat API contract while shrinking the spike to `/chat/completions`, `/health`, and `/summarize` only.

AIA: [PAI SeCeNc Hin R Curosr's GPT-5.4 1M Extra High · MAX v1.0](https://aiattribution.github.io/statements/AIA-PAI-SeCeNc-Hin-R-?model=Curosr%27s%20GPT%252D5.4%201M%20Extra%20High%20%C2%B7%20MAX-v1.0)

File: agents/langgraph/templates/ci_failure_summarizer/main.py [hunk 12]
```diff
@@ -0,0 +1,650 @@
...
+    created: int = Field(
+        ...,
+        description="The Unix timestamp (in seconds) of when the chat completion was created.",
+    )
+    model: str = Field(..., description="The model used for the chat completion.")
+    choices: list[Choice] = Field(..., description="A list of chat completion choices.")
+
+    usage: dict | None = Field(
+        None, description="Usage statistics for the completion request."
+    )
```
[the remark this answers is not in the document]
Comment (kami619) [reply agents/langgraph/templates/ci_failure_summarizer/main.py:126]: Fixed in `40e8f94`. `ChatCompletionResponse` now includes `context`, and `tests/test_api_contract.py` has a FastAPI response-model regression test that verifies `/chat/completions` preserves the field instead of dropping it during serialization.

AIA: [PAI SeCeNc Hin R Curosr's GPT-5.4 1M Extra High · MAX v1.0](https://aiattribution.github.io/statements/AIA-PAI-SeCeNc-Hin-R-?model=Curosr%27s%20GPT%252D5.4%201M%20Extra%20High%20%C2%B7%20MAX-v1.0)

File: agents/langgraph/templates/ci_failure_summarizer/Makefile [hunk 13]
```diff
@@ -0,0 +1,183 @@
...
+	  [ -n "$${CONTAINER_IMAGE}" ] || { echo "ERROR: CONTAINER_IMAGE is not set in .env"; exit 1; } && \
+	  LAST_SEG="$${CONTAINER_IMAGE##*/}" && if [[ "$$LAST_SEG" == *:* ]]; then IMAGE_REPO="$${CONTAINER_IMAGE%:*}"; IMAGE_TAG="$${LAST_SEG##*:}"; else IMAGE_REPO="$${CONTAINER_IMAGE}"; IMAGE_TAG="latest"; fi && \
+	  trap 'rm -f .helm-secrets.yaml' EXIT && \
+	  umask 077 && \
+	  API_KEY="$${API_KEY}" POSTGRES_PASSWORD="$${POSTGRES_PASSWORD}" \
+	  MLFLOW_TRACKING_TOKEN="$${MLFLOW_TRACKING_TOKEN}" GITHUB_TOKEN="$${GITHUB_TOKEN}" \
+	  SLACK_WEBHOOK_URL="$${SLACK_WEBHOOK_URL}" python3 -c 'import json, os, pathlib; doc={"secrets":{"apiKey": os.environ["API_KEY"], "postgresPassword": os.environ["POSTGRES_PASSWORD"]}}; mlflow=os.environ.get("MLFLOW_TRACKING_TOKEN"); extra={k: os.environ[k] for k in ("GITHUB_TOKEN", "SLACK_WEBHOOK_URL") if os.environ.get(k)}; \
+if mlflow: doc["secrets"]["mlflowTrackingToken"]=mlflow; \
+if extra: doc["extraSecrets"]=extra; \
+pathlib.Path(".helm-secrets.yaml").write_text(json.dumps(doc) + "\n")' && \
```
[the remark this answers is not in the document]
Comment (kami619) [reply agents/langgraph/templates/ci_failure_summarizer/Makefile]: Fixed in `40e8f94`. The inline `python3 -c` deploy snippet now uses expression-based updates instead of semicolon-chained `if` statements, so `.helm-secrets.yaml` is written before Helm runs. I also added `tests/test_makefile.py` to compile-check the embedded Python snippet.

AIA: [PAI SeCeNc Hin R Curosr's GPT-5.4 1M Extra High · MAX v1.0](https://aiattribution.github.io/statements/AIA-PAI-SeCeNc-Hin-R-?model=Curosr%27s%20GPT%252D5.4%201M%20Extra%20High%20%C2%B7%20MAX-v1.0)

File: agents/langgraph/templates/ci_failure_summarizer/README.md [hunk 14]
```diff
@@ -0,0 +1,575 @@
...
+oc get route langgraph-ci-failure-summarizer-agent -o jsonpath='{.spec.host}'
+```
+
+#### Remove deployment (`make undeploy`)
+
+```bash
+make undeploy
+```
+
+See [OpenShift Deployment](../../../docs/openshift-deployment.md) for more details.
```
[the remark this answers is not in the document]
Comment (kami619) [reply agents/langgraph/templates/ci_failure_summarizer/README.md]: Fixed in `40e8f94` by changing the OpenShift deployment reference to `../../../../docs/openshift-deployment.md` and verifying the resolved path against the repo layout.

AIA: [PAI SeCeNc Hin R Curosr's GPT-5.4 1M Extra High · MAX v1.0](https://aiattribution.github.io/statements/AIA-PAI-SeCeNc-Hin-R-?model=Curosr%27s%20GPT%252D5.4%201M%20Extra%20High%20%C2%B7%20MAX-v1.0)

File: agents/langgraph/templates/ci_failure_summarizer/src/ci_failure_summarizer/incident_store.py [hunk 15]
```diff
@@ -0,0 +1,273 @@
...
+    production would want a shared pool or context-managed lifecycle.
+    """
+
+    def __init__(self, db_uri: str) -> None:
+        self.db_uri = db_uri
+
+    @contextmanager
+    def _connection(self) -> Iterator[psycopg.Connection]:
+        with psycopg.connect(self.db_uri, row_factory=dict_row) as conn:
+            yield conn
```
[the remark this answers is not in the document]
Comment (kami619) [reply agents/langgraph/templates/ci_failure_summarizer/src/ci_failure_summarizer/incident_store.py:75]: Fixed in `40e8f94`. `IncidentStore._connection()` now passes `connect_timeout=10`, and `tests/test_incident_store.py` covers both the timeout wiring and the surrounding summary/claim branches that were missing direct coverage.

AIA: [PAI SeCeNc Hin R Curosr's GPT-5.4 1M Extra High · MAX v1.0](https://aiattribution.github.io/statements/AIA-PAI-SeCeNc-Hin-R-?model=Curosr%27s%20GPT%252D5.4%201M%20Extra%20High%20%C2%B7%20MAX-v1.0)

File: agents/langgraph/templates/ci_failure_summarizer/README.md [hunk 16]
```diff
@@ -0,0 +1,575 @@
...
+API_KEY = your-api-key-here
+BASE_URL = https://your-model-endpoint.com/v1
+MODEL_ID = llama-3.1-8b-instruct
+CONTAINER_IMAGE = quay.io/your-username/langgraph-ci-failure-summarizer-agent:latest
+
+POSTGRES_HOST = your-postgres-host.com
+POSTGRES_PORT = 5432
+POSTGRES_DB = agent_memory
+POSTGRES_USER = your_db_user
+POSTGRES_PASSWORD = your_db_password
```
[the remark this answers is not in the document]
Comment (kami619) [reply agents/langgraph/templates/ci_failure_summarizer/README.md]: Fixed in `54fdfa0`. The sample `.env` block in the README now uses shell-compatible `KEY=value` assignments for `API_KEY`, `BASE_URL`, `MODEL_ID`, `CONTAINER_IMAGE`, and the `POSTGRES_*` variables so `source .env` works as documented.

AIA: [PAI SeCeNc Hin R Curosr's GPT-5.4 1M Extra High · MAX v1.0](https://aiattribution.github.io/statements/AIA-PAI-SeCeNc-Hin-R-?model=Curosr%27s%20GPT%252D5.4%201M%20Extra%20High%20%C2%B7%20MAX-v1.0)

File: agents/langgraph/templates/ci_failure_summarizer/README.md [hunk 17]
```diff
@@ -0,0 +1,575 @@
...
+
+```bash
+oc login -u "login" -p "password" https://super-link-to-cluster:111
+```
+
+Login ex. Docker
+
+```bash
+docker login -u='login' -p='password' quay.io
+```
```
[the remark this answers is not in the document]
Comment (kami619) [reply agents/langgraph/templates/ci_failure_summarizer/README.md:363]: Fixed in `54fdfa0`. The README now uses interactive `oc login` and a `--password-stdin` Docker example so credentials are no longer shown directly on the command line.

AIA: [PAI SeCeNc Hin R Curosr's GPT-5.4 1M Extra High · MAX v1.0](https://aiattribution.github.io/statements/AIA-PAI-SeCeNc-Hin-R-?model=Curosr%27s%20GPT%252D5.4%201M%20Extra%20High%20%C2%B7%20MAX-v1.0)

**Comment by kami619** [comment]
@mpk-droid follow-up on the scope comments: the latest cleanup in `e3a1de1` addresses the four concerns directly. I added dedicated `tests/test_failure_evidence.py` coverage for the extraction, ranking, signature, and resolution helpers; removed the inherited playground UI and its packaging/test surface; dropped the chat-focused `tests/behavioral/` suite; and rewrote the summarizer README so `/summarize` is explicitly called out as deterministic while `MODEL_ID`/`BASE_URL`/`API_KEY` are documented as inherited `/chat/completions` scaffold requirements rather than part of the summarization path.

AIA: [PAI SeCeNc Hin R Curosr's GPT-5.4 1M Extra High · MAX v1.0](https://aiattribution.github.io/statements/AIA-PAI-SeCeNc-Hin-R-?model=Curosr%27s%20GPT%252D5.4%201M%20Extra%20High%20%C2%B7%20MAX-v1.0)

File: agents/langgraph/templates/ci_failure_summarizer/src/ci_failure_summarizer/github_client.py (hunk 2 above)
Comment (kami619) [reply agents/langgraph/templates/ci_failure_summarizer/src/ci_failure_summarizer/github_client.py:178]: Fixed in `6eca6ce`. `GitHubActionsClient.list_jobs()` now emits a warning when the returned job count meets the `per_page` boundary, so silent truncation becomes visible before QG4 outgrows the current single-page assumption. I kept the code path intentionally small for the spike rather than implementing full pagination in the same change.

AIA: [PAI SeCeNc Hin R Curosr's GPT-5.4 1M Extra High · MAX v1.0](https://aiattribution.github.io/statements/AIA-PAI-SeCeNc-Hin-R-?model=Curosr%27s%20GPT%252D5.4%201M%20Extra%20High%20%C2%B7%20MAX-v1.0)

File: agents/langgraph/templates/ci_failure_summarizer/src/ci_failure_summarizer/github_client.py (hunk 3 above)
Comment (kami619) [reply agents/langgraph/templates/ci_failure_summarizer/src/ci_failure_summarizer/github_client.py:169]: Follow-up in `6eca6ce`: I left the broader push-model redesign out of scope for this spike, but I updated the summarizer README to document the current assumption explicitly. The docs now call out that latest-run selection is single-page/single-latest today, and recommend using an explicit `run_id` (or a future push-style workflow callback) when multiple relevant runs may occur between trigger executions.

AIA: [PAI SeCeNc Hin R Curosr's GPT-5.4 1M Extra High · MAX v1.0](https://aiattribution.github.io/statements/AIA-PAI-SeCeNc-Hin-R-?model=Curosr%27s%20GPT%252D5.4%201M%20Extra%20High%20%C2%B7%20MAX-v1.0)

File: agents/langgraph/templates/ci_failure_summarizer/src/ci_failure_summarizer/github_client.py (hunk 4 above)
Comment (kami619) [reply agents/langgraph/templates/ci_failure_summarizer/src/ci_failure_summarizer/github_client.py:107]: Fixed the HTTP-session part in `6eca6ce`. `GitHubActionsClient` now has a `close()` method and `SummarizerOrchestrator.run()` closes the per-run GitHub session in a `finally` block so `/summarize` does not leave the `requests.Session` hanging. I left PostgreSQL pooling unchanged because the spike already documents per-call `IncidentStore` connections as a conscious trade-off rather than a production-ready pool design.

AIA: [PAI SeCeNc Hin R Curosr's GPT-5.4 1M Extra High · MAX v1.0](https://aiattribution.github.io/statements/AIA-PAI-SeCeNc-Hin-R-?model=Curosr%27s%20GPT%252D5.4%201M%20Extra%20High%20%C2%B7%20MAX-v1.0)

File: agents/langgraph/templates/ci_failure_summarizer/src/ci_failure_summarizer/slack_notifier.py (hunk 5 above)
Comment (kami619) [reply agents/langgraph/templates/ci_failure_summarizer/src/ci_failure_summarizer/slack_notifier.py]: Fixed in `6eca6ce`. `post_summary()` now redacts webhook URLs from `response.text` before raising, and `tests/test_slack_notifier.py` has a regression test that verifies the webhook token does not leak back through the raised error path.

AIA: [PAI SeCeNc Hin R Curosr's GPT-5.4 1M Extra High · MAX v1.0](https://aiattribution.github.io/statements/AIA-PAI-SeCeNc-Hin-R-?model=Curosr%27s%20GPT%252D5.4%201M%20Extra%20High%20%C2%B7%20MAX-v1.0)

**Review by mpk-droid — APPROVED** [review]
lgtm
